core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -299,41 +300,55 @@ public boolean topicExists(final String topic) throws TopicRepositoryException {
-                    .collect(
-                            Collectors.toMap(
-                                    p -> String.valueOf(p.partition()),
-                                    p -> p.leader().idString() + "_" + p.leader().host()));
             batch.forEach(item -> {
                 Preconditions.checkNotNull(
                         item.getPartition(), "BatchItem partition can't be null at the moment of publishing!");
-                item.setBrokerId(partitionToBroker.get(item.getPartition()));
             });
 
-            final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
+            final Function<BatchItem, CompletableFuture<Exception>> sendItem =
+                    item -> publishItem(producer, topicId, eventType, item, delete);
+
+            final List<CompletableFuture<Exception>> futures = new LinkedList<>();
+            final Map<String, CompletableFuture<Exception>> futuresByKey = new HashMap<>();
+
             for (final BatchItem item : batch) {
-                final String brokerId = item.getBrokerId();
-                if (brokerId == null) {
-                    item.updateStatusAndDetail(EventPublishingStatus.FAILED,
-                            String.format("No leader for partition: %s, topic: %s.", item.getPartition(), topicId));
-                    LOG.error("Failed to publish to kafka. No leader for ({}:{}).",
-                            topicId, item.getPartition());
-                    continue;
+                final String itemKey = item.getEventKey();
+                if (itemKey == null) {
+                    // items without a key go out of order
+                    futures.add(sendItem.apply(item));
+                } else {
+                    // items for a single key are composed together for sequential publishing
+                    CompletableFuture<Exception> fut = futuresByKey.get(itemKey);
+                    if (fut == null) {
+                        fut = sendItem.apply(item);
+                    } else {
+                        fut = fut.thenCompose(e -> {
+                                    if (e != null) {
+                                        item.updateStatusAndDetail(EventPublishingStatus.ABORTED,
+                                                "earlier item failed for key: " + itemKey);
+                                        return CompletableFuture.completedStage(e);
+                                    } else {
+                                        return sendItem.apply(item);
+                                    }
+                                });
+                    }
+                    futuresByKey.put(itemKey, fut);
                 }
-                item.setStep(EventPublishingStep.PUBLISHING);
-                sendFutures.put(item, publishItem(producer, topicId, eventType, item, delete));
             }
+            // we will wait for all futures together
```

Comment:
Btw, there is now enough code shared between the JSON and binary send methods, to think about extracting it into a common method, IMO.